### ISSUE TITLE
[1.8] Villager Registry seems to be incomplete

### ISSUE THREAD

**Issue by 61352151511** [MinecraftForge/MinecraftForge#1687]
It seems the villager registry is incomplete, registering custom villager ID's and professions appears to have no effect.

VillagerRegistry.instance().registerVillagerId(ConfigurationHandler.getVillagerID()); // (65)
Does keep track of the fact that the ID is registered.
VillagerRegistry.instance().registerVillagerSkin(65, new ResourceLocation("zombieinfection", "textures/entity/medic.png"));
Not sure if this works, on the client proxy, any villager spawned from a spawn egg does not have the skin.
VillagerProfession prof = new VillagerProfession("zombieinfection:medic", "zombieinfection:textures/entity/medic.png");
VillagerRegistry.instance().register(prof);
Again not sure if this works, no villagers seem to have the profession or skin. Also can't figure out how one would add custom trades, the .init method for VillagerCareer's takes ITradeList[][] however .init is hidden/not visible.

There's two TODO's in VillagerRegistry, not sure if that is what prevents the entire thing from working.

    public static void setRandomProfession(EntityVillager entity, Random rand)
    {
        //TODO: Grab id range from internal registry
        entity.setProfession(rand.nextInt(5));
    }

    //TODO: Figure out a good generic system for this. Put on hold for Patches.

If this doesn't belong here (Since it may be FML) I'll repost it there.

**Comment by LexManos** [comment]
Ah, totally forgot about the re-write of the villager registry u.u
It's completely changed in 1.8. Need to sit down and finish it.

**Comment by Chimaine** [comment]
I actually wondered about that when I looked for updating to 1.8. Thought it was intentional. Glad it's not and just want to bump this a bit, because it's one of the things preventing me to upgrade.

**Comment by cpw** [comment]
It's being worked on. Trying to make something more general purpose because it's not just villagers and their ids that need network sync and persistence..

**Comment by Parker8283** [comment]
Why, cpw! Good to see you again!

**Comment by tyronx** [comment]
Completion of this open issue would be great! I also stumbled upon this when upgrading my mod to 1.8

**Comment by tyronx** [comment]
Hm yea, unable to update all parts of my mod to 1.8 due to the missing vilager registry. I hope it's still being worked on.

**Comment by durand1w** [comment]
Just wanted to see if there was any news or progress on rolling out a villager registry so we can make plans for our 1.8 update.

**Comment by LexManos** [comment]
Its mostly in place, but the key aspect i've been asking people to provide, a test mod, has yet to be done. And i've been out of the country and then sick as a dog since the 28th so I'll get to it eventually.

**Comment by durand1w** [comment]
Sorry to hear about you being sick.  That's never fun.

Thanks for the update.  I wasn't sure that things were even that far along.  Thank you!

**Comment by rubensworks** [comment]
@LexManos Do you need a dummy test mod just for villagers? Or a complete mod for 1.8 that requires custom villagers? If the latter, I have a 1.8 mod available that only requires villagers to be implemented for it to be complete, so I could help you with that.

**Comment by LexManos** [comment]
Why would I need something that does 1000 different things to test 1 specific thing?

**Comment by dvatreknerd314** [comment]
Has there been any progress on this issue?

**Comment by LexManos** [comment]
Nobody has stepped up like I asked so I figured that nobody gave a shit.
I'm working on other more important things until someones cares.

**Comment by dvatreknerd314** [comment]
I have a mod that, so far, adds a custom village building and six items.  I don't yet have a texture for the villager created, but I have some trading options in mind.  Would this alone be sufficient?

**Comment by cpw** [comment]
If you wish to make village registry work, we'll welcome a pull request.. I don't believe there's much if anything to do..

**Comment by dvatreknerd314** [comment]
I seem to be having problems with GitHub, namely that I can't seem to clone stuff to my desktop.  I also can't find the Village Registry file in the GitHub repository

**Comment by dvatreknerd314** [comment]
It was also my understanding that what was needed to get the VillageRegistry working was test mods, and that's what I was trying to provide

**Comment by dvatreknerd314** [comment]
I'm not super familiar with either GitHub or Forge to know what all to do to get the VillagerRegistry to work.  I'm seeing some stuff in the code but I'm not sure how to modify either forge or vanilla code.

**Comment by durand1w** [comment]
Honeslty, I was hoping someone with more experience and familiarity with Forge would offer.  What exactly do you need out of a 1.8 mod to test the villager registry successfully?

**Comment by tyronx** [comment]
What LexManos probably wants is a test suite that covers as many VillagerRegistry use cases as possible made from a near zero documentation so he can re-implement something that already existed in 1.7.10 whose implementation for whatever reason didn't require a test mod....(or that test mod magically disappeared)

**Comment by williewillus** [comment]
iirc it's because vanilla made some background changes to villagers in 1.8 so a new registry needs to be made and hooked

**Comment by gabizou** [comment]
I'm just assuming here as to how the `VillagerRegistry` is **intended** to be used, though I'm not sure. I'm providing a simple mod code for testing, though correct me on my usage.

```java
package com.gabizou;

import net.minecraft.entity.passive.EntityVillager;
import net.minecraftforge.event.entity.EntityEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.VillagerRegistry;

@Mod(modid = "com.gabizou.VillagerTest", name = "SimpleVillagerTest", version = "1.0")
public class SimpleVillagerTest {

    public static final VillagerRegistry.VillagerProfession PROFESSION = new VillagerRegistry.VillagerProfession("Derpington", "minecraft:textures/entity/villager/farmer.png");
    public static final VillagerRegistry.VillagerCareer CAREER = new VillagerRegistry.VillagerCareer(PROFESSION, "Derpington");

    @SubscribeEvent
    public void onInit(FMLInitializationEvent event) {
        VillagerRegistry.instance().register(PROFESSION);
    }

    @SubscribeEvent
    public void spawning(EntityEvent.EntityConstructing entityConstructing) {
        if (entityConstructing.entity instanceof EntityVillager) {
            EntityVillager villager = (EntityVillager) entityConstructing.entity;
            villager.setProfession(PROFESSION.getId()); // This is likely how it should be used? I'm not 100% sure...
            villager.setCareer(CAREER); // Again, how it should be used likely?
        }
    }

}

```

[A gist was made as well](https://gist.github.com/gabizou/ae9b2807a0cf7e0cdeb9)

**Comment by 61352151511** [comment]
The thing is, In 1.7.10 we did not need to manually set the profession or career whenever a villager spawns. Also that would set every villager that spawns to your villager, so unless every time a mod adds a villager they'd have to do that and have a RNG determine whether to switch it to their villager or not.

**Comment by LexManos** [comment]
Minecraft sets the profession not you.
Sadly it does it randomly but meh. {This needs to be fully implemented once the registry is complete, as in when someone cares to make a test mod}

**Comment by gabizou** [comment]
> Also that would set every villager that spawns to your villager

That's the entire point of the example. 

> when someone cares to make a test mod

If this isn't a test mod, then can you explain what else the mod needs to have?

**Comment by Thutmose** [comment]
the only thing I see missing would be a way to set the trades for the career 

```java
package test.villagertest;

import net.minecraft.entity.passive.EntityVillager;
import net.minecraft.entity.passive.EntityVillager.EmeraldForItems;
import net.minecraft.entity.passive.EntityVillager.PriceInfo;
import net.minecraft.init.Blocks;
import net.minecraft.item.Item;
import net.minecraftforge.event.entity.EntityEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.VillagerRegistry;

@Mod(modid = "VillagerTest", name = "SimpleVillagerTest", version = "1.0")
public class SimpleVillagerTest {

    public static final VillagerRegistry.VillagerProfession PROFESSION = new VillagerRegistry.VillagerProfession("Derpington", "minecraft:textures/entity/villager/farmer.png");
    public static final VillagerRegistry.VillagerCareer CAREER = new VillagerRegistry.VillagerCareer(PROFESSION, "Derpington");

    @SubscribeEvent
    public void onInit(FMLInitializationEvent event) {
        VillagerRegistry.instance().register(PROFESSION);
        //add a trade for the career, using an addTrade(ITradeList trade) method in VillagerCareer
        CAREER.addTrade(new EmeraldForItems(Item.getItemFromBlock(Blocks.dirt), new PriceInfo(1, 1)));
        
    }

    @SubscribeEvent
    public void spawning(EntityEvent.EntityConstructing entityConstructing) {
    	//Set all spawning villagers to new profession for a test, 
    	//in reality, these methods will be used when the villager is constructed elsewhere for spawning.
        if (entityConstructing.entity instanceof EntityVillager) {
            EntityVillager villager = (EntityVillager) entityConstructing.entity;
            
            /*
             * This VillagerProfession.getId() should probably check a value set while regsitering the profession
             * with VillagerRegistry.instance().register(PROFESSION);
             */
            villager.setProfession(PROFESSION.getId());
            
            /**
             * If this Method isn't called, it should randomly select a career from the list of careers for that
             * profession, every time new VillagerRegistry.VillagerCareer(PROFESSION, "Derpington"); is called,
             * it should add that new VillageCareer to a list for PROFESSION
             */
            villager.setCareer(CAREER);
        }
    }
}
```

here is a gist link: https://gist.github.com/Thutmose/1b11be61a4d340fa45ae

**Comment by 61352151511** [comment]
@Thutmose 

> Minecraft sets the profession not you.
> Sadly it does it randomly but meh. {This needs to be fully implemented once the registry is complete, as in when someone cares to make a test mod}

Basically what he is saying (From what I understand) is that the villager registry is not fully complete. Someone needs to make a pull request and complete the villager registry. In order for it to be considered "complete" Minecraft itself should be able to spawn the villagers when a village is generated, you shouldn't have to hook into an event and set it yourself.

In 1.7.10 You would register a villager ID with "VillagerRegistry.instance().registerVillagerId(intVillagerId)" then a trade handler with "VillagerRegistry.instance().registerVillageTradeHandler(intVillagerId, new TradeHandler())" (TradeHandler is a class implementing IVillageTradeHandler) I think you needed a bit more but it was very simple to do and you wouldn't need to use an event to change the entity that minecraft is spawning.

If I had the time I'd look into making a villager registry myself, however I don't have any at the moment and probably won't for a long while so if someone wants to task themself with making the 1.8 villager registry more 1.7.10ish they should.

**Comment by Thutmose** [comment]
so in vanilla, when a villager spawns in randomly, IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, IEntityLivingData livingdata) is called, which then calls net.minecraftforge.fml.common.registry.VillagerRegistry.setRandomProfession(this, this.worldObj.rand); which already handles the randomization, and seems to already be planned to add proper randomization based on internal registry.

The point of that mod was to demonstrate the use of being able to manually set profession and career by a mod, this simple mod does it to all villagers, as it is just a testing mod, It would not normally be done during that event used, but rather whenever the mod needs to spawn in the villager, by calling the methods in the event, after it has constructed the villager.

The changes to the EntityVillager class should probably be as follows (some of these are planned already): 
Use villager registry for the trade list map, or just remove it/ignore it entirely
remove the % 5 in the getProfession(), optinally replace it with number of professions registered.
populateBuyingList() would use VillagerRegistry. (making trade list map obsolete)
getDisplayName() should check VillagerRegistry.

Currently career is randomly assigned (line 596 of the EntityVillager.class I am looking at) but changing careerId to public, or assigning setter for it, would allow that to be changed.

**Comment by LexManos** [comment]
The issue that I was running into when dealing with the reigstry is that the NameSpacedRegistry is not ordered.
So entry 1 may not have id 1.
And the way vanilla placed villagers was using rand.nextInt(5) in the world gen. And we need to keep vanilla world-gen compatibility so i've been forced to keep that rand.nextInt(5).
There may be a way to say screw it and fork the world gen at that point IF there are extra villagers registered. {I think we do this in a few other world-gen areas} But I needed someone to make me a villager with some way of identifying that villager so I can test.
Or if someone wants to look into it for me.

Feel free to start actually looking into it and making a PR. Make sure you compare with vanilla generation because keeping that compatibility is very important. God I wish that Mojang would move away from a integer id system and use strings.

**Comment by Thutmose** [comment]
ok, I will see if I can get something working, idk if what I write will be correctly formatted, but I should be able to get something to work, and then I will see about trying to get it to the proper style standard.

**Comment by williewillus** [comment]
6e04a6c

**Comment by Actuarius** [comment]
@williewillus closed this issue
